File: kotlin/src/main/kotlin/advertisement/infrastructure/persisence/SqLiteAdvertisementRepository.kt
package advertisement.infrastructure.persisence

import advertisement.domain.AdvertisementRepository
import advertisement.domain.model.Advertisement
import framework.database.DatabaseConnection
import java.security.MessageDigest

class SqLiteAdvertisementRepository(private val connection: DatabaseConnection): AdvertisementRepository {
    override fun save(advertisement: Advertisement) {
        val passwordHash = advertisement.password!!.md5()
        connection.execute(
            "INSERT INTO advertisements (id, description, password) VALUES ('" +
                    "${advertisement.id}', '${advertisement.description}', '$passwordHash')"
        )
    }

    private fun String.md5(): String {
        val md = MessageDigest.getInstance("MD5")
        val digest = md.digest(this.toByteArray())
        return digest.toString()
    }
}

File: typescript/src/advertisement/infrastructure/SqliteAdvertisementRepository.ts
import { AdvertisementRepository } from '../domain/AdvertisementRepository';
import { Advertisement } from '../domain/model/Advertisement';
import { createHash } from "node:crypto";
import { DatabaseConnection } from '../../framework/database/DatabaseConnection';

export class SqliteAdvertisementRepository implements AdvertisementRepository {

  constructor(
    private connection: DatabaseConnection) {
  }

  async save(advertisement: Advertisement): Promise<void> {

    await this.connection.execute(
      `INSERT INTO advertisements (id, description, password) 
      VALUES (?, ?, ?) 
      ON CONFLICT(id) DO UPDATE 
      SET description = excluded.description, password = excluded.password`, [advertisement.id(), advertisement.description(), createHash('md5').update(advertisement.password()).digest('hex')]);

  }
}
